kube-runtime/src/reflector/store.rs

```diff
@@ -105,39 +108,66 @@ where
                 let key = obj.to_object_ref(self.dyntype.clone());
                 self.store.write().remove(&key);
             }
-            watcher::Event::Restarted(new_objs) => {
+            watcher::Event::RestartedStart => {
+                let mut buffer = self.buffer.write();
+                *buffer = AHashMap::new();
+            }
+            watcher::Event::RestartedPage(new_objs) => {
                 let new_objs = new_objs
                     .iter()
                     .map(|obj| (obj.to_object_ref(self.dyntype.clone()), Arc::new(obj.clone())))
                     .collect::<AHashMap<_, _>>();
-                *self.store.write() = new_objs;
+                self.buffer.write().extend(new_objs);
```

Comment:
thinking out loud: it's probably still cheaper to collect into a ahashmap and do a one step `extend` (like this), than to avoid the temporary and do multiple inserts with the lock / multiple lock acquires. but maybe that changes if we can avoid the lock on the `buffer`.